test/parallel_api/iterator/zip_iterator.pass.cpp

```diff
@@ -121,6 +121,10 @@ DEFINE_TEST(test_for_each)
 
         auto tuple_first1 = oneapi::dpl::make_zip_iterator(std::make_tuple(first1, first1));
         auto tuple_last1 = oneapi::dpl::make_zip_iterator(last1, last1);
+        if (!this->host_buffering_required())
```

Comment:
@danhoeflinger, could you please explain this check, why it's required?

Reply:
This is required to avoid fancy iterators which are composed of `sycl_iterator` types.  We don't expect a `zip_iterator` of `sycl_iterators` to be "device copyable", because  

"the copy constructor selected for every non-static class type (or array of class type) member of T is trivial;" 
is no longer true from https://en.cppreference.com/w/cpp/language/copy_constructor#Trivial_copy_constructor.  `zip_iterator` contains a tuple of `sycl_iterators` which are not "trivial".

According to https://en.cppreference.com/w/cpp/language/classes#Trivially_copyable_class
"each eligible copy constructor is trivial" must be true for a type to be trivially copyable.

This is fine, because with the design of `__get_sycl_range`, we don't expect such a type to be directly device_copyable.  Once such a type is processed by `__get_sycl_range` it ends up as a `zip_view<all_view, all_view>` where all_view contains a device_copyable `sycl::accessor`.  This resulting type is device copyable.

The current design only requires fancy iterators which are made up of fully "passed_directly" components to be device copyable.

Reply:
Sure, we should be able to check `is_passed_directly` instead.  Thanks.

Reply:
Actually, considering this further, we would be reaching into private API to check either `is_passed_directly` or `is_sycl_iterator`.  We could redundantly add these structs in testing utils, but we already have the same information in this function (the point of this function is to determine if the data is in a host data destined for a sycl::buffer or USM memory).

I prefer to add a comment and keep this call as is.